### PR TITLE
Parse turbofish

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: ast/src/parsed/mod.rs [hunk 1]
```diff
@@ -475,11 +475,15 @@ pub struct PolynomialName {
 /// This is different from SymbolPath mainly due to different formatting.
 pub struct NamespacedPolynomialReference {
     pub path: SymbolPath,
+    pub generic_args: Option<Vec<Type<Expression>>>,
```
Comment (chriseth) [inline ast/src/parsed/mod.rs:478]: Maybe call this `type_args`?

File: pil-analyzer/src/statement_processor.rs [hunk 2]
```diff
@@ -512,19 +520,27 @@ where
...
 
-    fn process_expression(&self, expr: parsed::Expression) -> Expression {
-        self.expression_processor().process_expression(expr)
+    fn process_expression(
+        &self,
+        expr: parsed::Expression,
+        type_vars: &HashSet<&String>,
+    ) -> Expression {
+        self.expression_processor(type_vars)
+            .process_expression(expr)
```
Comment (pacheco) [inline pil-analyzer/src/statement_processor.rs:536]: nitpick: should there be wrapping methods for the other `.expression_processor().process_foo` calls?

File: pil-analyzer/src/type_inference.rs [hunk 3]
```diff
@@ -120,8 +122,14 @@ impl TypeChecker {
 
             let declared_type = self.declared_types[&name].clone();
             let result = if declared_type.vars.is_empty() {
+                self.declared_type_vars = Default::default();
```
Comment (pacheco) [inline pil-analyzer/src/type_inference.rs:125]: `self.declared_type_vars.clear()`?

File: parser/src/powdr.lalrpop [hunk 4]
```diff
@@ -60,6 +60,19 @@ Part: Part = {
     <name:Identifier> => Part::Named(name),
 }
 
+// Same as SymbolPath plus we allow "::<...>" at the end.
```
[the remark this answers is not in the document]
Comment (chriseth) [reply parser/src/powdr.lalrpop:63]: No because of the comment in line 65

File: pil-analyzer/src/statement_processor.rs (hunk 2 above)
Comment (chriseth) [reply pil-analyzer/src/statement_processor.rs:536]: Do you have an example? There are other wrapping methods as well, for example process_selected_expression.

File: pil-analyzer/tests/parse_display.rs [hunk 5]
```diff
@@ -52,6 +49,7 @@ namespace T(65536);
     T.A' = (((T.first_step' * 0) + (T.reg_write_X_A * T.X)) + ((1 - (T.first_step' + T.reg_write_X_A)) * T.A));
     col witness X_free_value(__i) query match std::prover::eval(T.pc) {
         0 => std::prover::Query::Input(1),
+        3 => std::prover::Query::Input(std::convert::int::<fe>((std::prover::eval(T.CNT) + 1))),
```
[the remark this answers is not in the document]
Comment (chriseth) [reply pil-analyzer/tests/parse_display.rs:52]: yeah, maybe. Looks like I'm still thinking in C++ terms. Once we have traits, we can also just use
``Into::into(...)`, which would then be made explicit as ``Into::into<fe, int>(...)``
(traits in pil don't have an implicit "Self" parameter)

File: pil-analyzer/src/statement_processor.rs (hunk 2 above)
Comment (chriseth) [reply pil-analyzer/src/statement_processor.rs:536]: Oh, I just realized I don't use them.

Comment (chriseth) [reply pil-analyzer/src/statement_processor.rs:536]: Ah no, it's other cases. Yeah, maybe I should also remove the helpers, it's kind of cluttery.

**Review by leonardoalt — APPROVED** [review]

**Review by leonardoalt — APPROVED** [review]

**Pull request merged**